[Canvas] Adopt smart pointers in WebGL utilities classes

csaavedra opened the pull request:
#### 2aec492f9190e75bcc920e716d463f526f5508e9
<pre>
[Canvas] Adopt smart pointers in WebGL utilities classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=280619">https://bugs.webkit.org/show_bug.cgi?id=280619</a>

Reviewed by Chris Dumez.

Use smart pointers/references in these utility classes to address
uncounted member warnings from the static analyzer.

* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters):
(WebCore::ScopedDisableRasterizerDiscard::ScopedDisableRasterizerDiscard):
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer):
(WebCore::ScopedDisableScissorTest::ScopedDisableScissorTest):
(WebCore::ScopedWebGLRestoreFramebuffer::~ScopedWebGLRestoreFramebuffer):
(WebCore::ScopedWebGLRestoreRenderbuffer::~ScopedWebGLRestoreRenderbuffer):
(WebCore::ScopedWebGLRestoreTexture::~ScopedWebGLRestoreTexture):
(WebCore::ScopedClearColorAndMask::ScopedClearColorAndMask):
(WebCore::ScopedClearColorAndMask::~ScopedClearColorAndMask):
(WebCore::ScopedClearDepthAndMask::ScopedClearDepthAndMask):
(WebCore::ScopedClearStencilAndMask::ScopedClearStencilAndMask):

Canonical link: <a href="https://commits.webkit.org/284701@main">https://commits.webkit.org/284701@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17ad8855ece07c78f4b58865cda43ba399d1c3a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70190 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74277 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21356 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72307 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21208 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55668 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14146 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73256 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36129 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41820 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17961 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19731 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18305 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76006 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14415 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17543 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63388 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60597 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63324 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15565 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11343 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4960 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45398 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/164 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46472 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47746 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46213 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -63,7 +63,7 @@ class ScopedInspectorShaderProgramHighlight {
     } m_savedBlend;
 
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
We can drop the `const` here.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -125,7 +125,7 @@ class ScopedDisableRasterizerDiscard {
 
 private:
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -320,7 +320,7 @@ class ScopedClearDepthAndMask {
 
 private:
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -348,7 +348,7 @@ class ScopedClearStencilAndMask {
 
 private:
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -103,7 +103,7 @@ class ScopedTightUnpackParameters {
             context->pixelStorei(GraphicsContextGL::UNPACK_SKIP_IMAGES, newValues.skipImages);
     }
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -155,7 +155,7 @@ class ScopedEnableBackbuffer {
 
 private:
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -177,7 +177,7 @@ class ScopedDisableScissorTest {
 
 private:
-    WebGLRenderingContextBase* const m_context;
+    WeakPtr<WebGLRenderingContextBase> const m_context;
```
ditto.

cdumez reviewed (approved):
LGTM with nits.

csaavedra committed:
[Canvas] Adopt smart pointers in WebGL utilities classes
https://bugs.webkit.org/show_bug.cgi?id=280619

Reviewed by Chris Dumez.

Use smart pointers/references in these utility classes to address
uncounted member warnings from the static analyzer.

* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters):
(WebCore::ScopedDisableRasterizerDiscard::ScopedDisableRasterizerDiscard):
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer):
(WebCore::ScopedDisableScissorTest::ScopedDisableScissorTest):
(WebCore::ScopedWebGLRestoreFramebuffer::~ScopedWebGLRestoreFramebuffer):
(WebCore::ScopedWebGLRestoreRenderbuffer::~ScopedWebGLRestoreRenderbuffer):
(WebCore::ScopedWebGLRestoreTexture::~ScopedWebGLRestoreTexture):
(WebCore::ScopedClearColorAndMask::ScopedClearColorAndMask):
(WebCore::ScopedClearColorAndMask::~ScopedClearColorAndMask):
(WebCore::ScopedClearDepthAndMask::ScopedClearDepthAndMask):
(WebCore::ScopedClearStencilAndMask::ScopedClearStencilAndMask):

Canonical link: https://commits.webkit.org/284701@main
